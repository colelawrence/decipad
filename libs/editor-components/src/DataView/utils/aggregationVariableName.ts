import { zip } from 'lodash';
import {
  type SerializedType,
  type Result,
  type Value,
  identifierRegExpGlobal,
} from '@decipad/remote-computer';
import { labelize } from '@decipad/parse';
import { getDefined } from '@decipad/utils';

// extracts the fixColumnName from the columnName e.g. 'year_1' -> 'year'
const fixColumnName = (columnName?: string): string => {
  if (!columnName) {
    return 'Unknown';
  }
  const match = columnName.match(/.*(_[0-9]+)$/);
  if (match && match[1]) {
    return columnName.slice(0, columnName.indexOf(match[1]));
  }
  return columnName;
};

export const aggregationVariableName = (
  columns: Array<Value.TreeColumn>,
  columnTypes: Array<SerializedType>,
  valuePath: Array<Result.OneResult>
): string | undefined => {
  try {
    const fixedColumnNames = columns.filter(Boolean).map((col) => {
      return fixColumnName(col.name);
    });

    const zipped = zip(fixedColumnNames, columnTypes, valuePath);
    const variableName = zipped.reduce(
      (expr, [columnName, columnType, value], iter) => {
        const fixedColumnName = fixColumnName(columnName);
        if (value != null && iter < fixedColumnNames.length - 1) {
          return expr === ''
            ? `${fixedColumnName}${labelize(value, getDefined(columnType))}`
            : `${expr}${fixedColumnName}${labelize(
                value,
                getDefined(columnType)
              )}`;
        } else if (value == null && iter === fixedColumnNames.length - 1) {
          return expr === '' ? fixedColumnName : `${fixedColumnName}${expr}`;
        }
        return expr;
      },
      ''
    ); // started with tableName but was too long
    return (
      variableName.match(new RegExp(identifierRegExpGlobal))?.join('') || ''
    );
  } catch (e) {
    console.error(e);
    return undefined;
  }
};
