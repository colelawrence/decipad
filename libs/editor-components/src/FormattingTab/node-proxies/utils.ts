import {
  MyNode,
  MyEditor,
  MyElement,
  VariableDefinitionElement,
} from '@decipad/editor-types';
import { convertVariableDefInto } from '@decipad/editor-utils';
import {
  findNodePath,
  replaceNodeChildren,
  setNodes,
} from '@udecode/plate-common';
import { ColorSwatchOrAuto } from '../proxy-fields';

export const setNodeProperty = <N extends MyNode, K extends keyof N>(
  editor: MyEditor,
  node: N,
  key: K,
  value: unknown extends N[K] ? never : N[K]
) => {
  const path = findNodePath(editor, node);
  if (!path) return;
  setNodes(editor, { [key]: value }, { at: path });
};

export const setNodeText = (
  editor: MyEditor,
  node: MyElement,
  text: string
) => {
  const path = findNodePath(editor, node);
  if (!path) return;
  replaceNodeChildren(editor, {
    at: path,
    nodes: [{ text }],
  });
};

export const mapVariableProperties = (node: VariableDefinitionElement) => {
  const [caption, expression] = node.children;
  return {
    name: caption.children[0].text,
    variant: node.variant,
    color: (caption.color ?? 'auto') as ColorSwatchOrAuto,
    value: expression.children[0].text,
  };
};

export const variableActions = {
  setName: (node: VariableDefinitionElement, editor: MyEditor, name: string) =>
    setNodeText(editor, node.children[0], name),
  setVariant: (
    node: VariableDefinitionElement,
    editor: MyEditor,
    variant: VariableDefinitionElement['variant']
  ) => convertVariableDefInto(editor, node)(variant),
  setColor: (
    node: VariableDefinitionElement,
    editor: MyEditor,
    color: ColorSwatchOrAuto
  ) =>
    setNodeProperty(
      editor,
      node.children[0],
      'color',
      color === 'auto' ? undefined : color
    ),
  setValue: (
    node: VariableDefinitionElement,
    editor: MyEditor,
    value: string
  ) => setNodeText(editor, node.children[1], value),
};

const normalizeNumber = (value: string, allowNegative: boolean) => {
  let normalizedValue = value;

  // Remove all characters except digits and '.'
  normalizedValue = normalizedValue.replaceAll(/[^0-9.]/g, '');

  // Reinsert '-' at start of number
  if (allowNegative && value.startsWith('-')) {
    normalizedValue = `-${normalizedValue}`;
  }

  return normalizedValue;
};

export const normalizeRationalNumber = (value: string) =>
  normalizeNumber(value, true);

export const normalizeNonNegativeNumber = (value: string) =>
  normalizeNumber(value, false);
