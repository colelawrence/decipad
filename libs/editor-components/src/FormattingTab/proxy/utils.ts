import { MultipleNodeProxyProperty } from './types';

// Map implementation for the MultipleNodeProxyProperty functor
export const mapProperty = <T, U>(
  property: MultipleNodeProxyProperty<T>,
  f: (value: T) => U
): MultipleNodeProxyProperty<U> =>
  property === 'varies' ? property : { value: f(property.value) };

export const withDefault = <T>(
  property: MultipleNodeProxyProperty<T | null | undefined>,
  defaultValue: T
): MultipleNodeProxyProperty<T> =>
  mapProperty(property, (value) => value ?? defaultValue);
