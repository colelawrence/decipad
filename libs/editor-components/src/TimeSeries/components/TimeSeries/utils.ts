import { DataGroup } from 'libs/editor-components/src/DataView/types';

export function getChildAtLevel(
  child: DataGroup,
  index: number,
  level: number
) {
  let currentChild = child;

  for (let i = 0; i < level; i++) {
    if (
      !currentChild ||
      !currentChild.children ||
      currentChild.children[index] === undefined
    ) {
      return undefined; // Return undefined if the child at the specified index doesn't exist
    }
    currentChild = currentChild.children[index];
  }

  return currentChild;
}
