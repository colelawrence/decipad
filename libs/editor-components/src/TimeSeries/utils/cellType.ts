import { type SerializedType } from '@decipad/language-interfaces';

export const cellType = (type: SerializedType): SerializedType => {
  if (type.kind !== 'column') {
    throw new Error('expected type to be column');
  }
  return type.cellType;
};
