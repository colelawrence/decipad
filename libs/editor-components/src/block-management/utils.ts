import type { Computer } from '@decipad/computer-interfaces';
import type { MyElement } from '@decipad/editor-types';
import { ELEMENT_TABLE, ELEMENT_VARIABLE_DEF } from '@decipad/editor-types';
import { clone } from '@decipad/editor-utils';
import { nanoid } from 'nanoid';

const utils = {
  /**
   * If you need any extra business logic on clone,
   * add it onto this function,
   * otherwise the function `clone` is used.
   */
  async cloneProxy(computer: Computer, element: MyElement) {
    const newElement = await clone(computer, element);
    switch (newElement.type) {
      case ELEMENT_VARIABLE_DEF: {
        if (newElement.variant === 'dropdown') {
          for (const c of newElement.children[1].options) {
            c.id = nanoid();
          }
        }
        break;
      }
      case ELEMENT_TABLE: {
        //
        // Look for category values, and de-duplicate the IDs
        //

        for (const col of newElement.children[1].children) {
          if (col.categoryValues == null) {
            continue;
          }

          for (const category of col.categoryValues) {
            category.id = nanoid();
          }
        }
      }
    }
    return newElement;
  },
};

export default utils;
