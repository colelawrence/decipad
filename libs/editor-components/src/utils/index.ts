export * from './input';
export * from './plot';
export * from './slashCommands';
export * from './table';
