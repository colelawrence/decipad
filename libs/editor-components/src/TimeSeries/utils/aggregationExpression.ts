import { zip } from 'lodash';
import { textify } from '@decipad/parse';
import { getDefined } from '@decipad/utils';
import {
  type SerializedType,
  type Result,
  type Value,
} from '@decipad/remote-computer';
import { getAggregationById } from '@decipad/language-aggregations';
import {
  type TimeSeriesFilter,
  type TableCellType,
} from '@decipad/editor-types';
import { getFilterExpression } from '@decipad/language-filters';

const fixColumnName = (columnName?: string): string => {
  if (!columnName) {
    return 'Unknown';
  }
  const match = columnName.match(/.*(_[0-9]+)$/);
  if (match && match[1]) {
    return columnName.slice(0, columnName.indexOf(match[1]));
  }
  return columnName;
};

const isComplexExpression = (expr: string): boolean =>
  expr.includes('.') || expr.includes('(');
const maybeEncloseExpression = (expr: string): string =>
  isComplexExpression(expr) ? `(${expr})` : expr;

const maybeAggregateExpression = (
  expr: string,
  aggregationId: string | undefined
): string => {
  if (!aggregationId) {
    return expr;
  }
  const aggregation = getAggregationById(aggregationId);
  if (!aggregation) {
    return expr;
  }
  return aggregation.expression(expr, {
    sum: `sum(${expr})`,
  });
};

const maybeFilterExpression = (
  tableName: string,
  columnName: string | undefined,
  filter: TimeSeriesFilter | undefined,
  columnType: TableCellType
): string => {
  if (!filter || !columnName) {
    return tableName;
  }

  const filterExpression = getFilterExpression(
    filter.operation,
    `${tableName}.${columnName}`,
    filter.valueOrValues,
    columnType
  );

  if (!filterExpression) {
    return tableName;
  }

  return `filter(${tableName}, ${filterExpression})`;
};

export const aggregationExpression = (
  tableName: string,
  columns: Array<Value.TreeColumn>,
  columnTypes: Array<SerializedType>,
  valuePath: Array<Result.OneResult>,
  filters: Array<TimeSeriesFilter | undefined>,
  aggregationName: string | undefined,
  rounding: string | undefined,
  filter: TimeSeriesFilter | undefined
): string | undefined => {
  // console.log('aggregationExpression', {
  //   tableName,
  //   columns,
  //   columnTypes,
  //   valuePath,
  //   aggregationName,
  // });
  if (!aggregationName) {
    return undefined;
  }
  const fixedColumnNames = columns.map((col) => fixColumnName(col.name));

  const filteredTableExpression = zip(fixedColumnNames, [
    ...filters,
    filter,
  ]).reduce((expr, [columnName, columnFilter], iter) => {
    return maybeFilterExpression(
      expr,
      columnName,
      columnFilter,
      columnTypes[iter] as TableCellType
    );
  }, tableName);

  // filter
  const zipped = zip(fixedColumnNames, columnTypes, valuePath);
  const tableExpression = zipped.reduce(
    (expr, [columnName, columnType, value], iter) => {
      const fixedColumnName = fixColumnName(columnName);
      if (value != null) {
        const columnExpression = `${
          iter === 0 ? expr : `(${expr})`
        }.${fixedColumnName}`;

        return `filter(${expr}, ${columnExpression} == ${textify(
          value,
          getDefined(columnType)
        )})`;
      }
      return expr;
    },
    filteredTableExpression
  );

  const columnName = fixedColumnNames[fixedColumnNames.length - 1];
  const columnExpression = `${maybeEncloseExpression(
    tableExpression
  )}.${columnName}`;
  const roundedColumnExpression = rounding
    ? `round(${columnExpression}, ${rounding})`
    : columnExpression;

  return maybeAggregateExpression(roundedColumnExpression, aggregationName);
};
