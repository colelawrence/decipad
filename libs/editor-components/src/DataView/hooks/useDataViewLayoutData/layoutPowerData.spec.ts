import { expect, describe, it } from 'vitest';
import { nanoid } from 'nanoid';
import { type Result, type AST, Unknown } from '@decipad/language-interfaces';
import type { IdentifiedBlock, Program } from '@decipad/computer-interfaces';
import { Value } from '@decipad/language-types';
import { getDefined } from '@decipad/utils';
import { buildResult, decilang, getComputer } from '@decipad/computer';
import { layoutPowerData } from './layoutPowerData';

const programBlock = (
  statement: AST.Statement,
  blockId = nanoid()
): IdentifiedBlock => {
  return {
    type: 'identified-block',
    id: blockId,
    block: {
      id: blockId,
      type: 'block',
      args: [statement],
    },
  };
};

const toProgram = (code: AST.Statement[]): Program => {
  return code.map((c, index) => programBlock(c, `block-${index}`));
};

describe('layoutPowerData', () => {
  it('works with an empty tree', async () => {
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree: buildResult(
          {
            kind: 'tree',
            columnTypes: [],
            columnNames: [],
          },
          Value.Tree.from(Unknown, undefined, [], [])
        ),
        aggregationTypes: [],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`[]`);
  });

  it('works with one column one group tree', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A"]
        }`,
          decilang`Tree = tree(DataTable)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-1']?.result,
      'no result for block-1'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchInlineSnapshot(`
      {
        "meta": undefined,
        "type": {
          "columnNames": [
            "Column1",
          ],
          "columnTypes": [
            {
              "kind": "string",
            },
          ],
          "kind": "tree",
        },
        "value": Tree {
          "children": [
            Tree {
              "children": [],
              "columns": [],
              "originalCardinality": 1,
              "root": "A",
              "rootAggregation": undefined,
            },
          ],
          "columns": [
            {
              "aggregation": undefined,
              "name": "Column1",
            },
          ],
          "originalCardinality": 1,
          "root": undefined,
          "rootAggregation": undefined,
        },
      }
    `);
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique'],
        expandedGroups: [],
        roundings: [],
        filters: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [],
          "collapsible": false,
          "elementType": "group",
          "id": "/a32fd8b31587b9c6354faa86e12756d2e7aca860346be182aacebb91f0b58026",
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [],
          "elementType": "smartrow",
          "global": true,
          "type": undefined,
          "value": Symbol(unknown),
        },
      ]
    `);
  });

  it('works with one column one group tree and aggregation', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A"]
        }`,
          decilang`Empty = {}`,
          decilang`Aggregations = {
          Column1 = 'string:count-unique'
        }`,
          decilang`Tree = tree(DataTable, Empty, Empty, Aggregations)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-3']?.result,
      'no result for block-3'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchInlineSnapshot(`
      {
        "meta": undefined,
        "type": {
          "columnNames": [
            "Column1",
          ],
          "columnTypes": [
            {
              "kind": "string",
            },
          ],
          "kind": "tree",
        },
        "value": Tree {
          "children": [
            Tree {
              "children": [],
              "columns": [],
              "originalCardinality": 1,
              "root": "A",
              "rootAggregation": {
                "meta": undefined,
                "type": {
                  "kind": "number",
                },
                "value": DeciNumber {
                  "d": 1n,
                  "infinite": false,
                  "n": 1n,
                  "s": 1n,
                },
              },
            },
          ],
          "columns": [
            {
              "aggregation": {
                "meta": undefined,
                "type": {
                  "kind": "number",
                },
                "value": DeciNumber {
                  "d": 1n,
                  "infinite": false,
                  "n": 1n,
                  "s": 1n,
                },
              },
              "name": "Column1",
            },
          ],
          "originalCardinality": 1,
          "root": undefined,
          "rootAggregation": undefined,
        },
      }
    `);
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique'],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [],
          "collapsible": false,
          "elementType": "group",
          "id": "/e85580031cb9166cd313930d684f09bdda1e3f129191a14419cc70382e41c685",
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [],
          "elementType": "smartrow",
          "global": true,
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 1n,
            "s": 1n,
          },
        },
      ]
    `);
  });

  it('works with two column two groups tree and aggregation', async () => {
    const computer = getComputer();
    const results = await computer.computeDeltaRequest({
      program: {
        upsert: toProgram([
          decilang`DataTable = {
          Column1 = ["A", "B", "A"]
          Column2 = [3, 2, 1]
        }`,
          decilang`Empty = {}`,
          decilang`Aggregations = {
            Column1 = 'string:count-unique'
            Column2 = 'number:sum'
          }`,
          decilang`Tree = tree(DataTable, Empty, Empty, Aggregations)`,
        ]),
      },
    });
    const tree = getDefined(
      results?.blockResults['block-3']?.result,
      'no result for block-3'
    ) as Result.Result<'tree'>;
    expect(tree).toMatchSnapshot();
    expect(
      await layoutPowerData({
        tableName: 'TableName',
        tree,
        aggregationTypes: ['string:count-unique', 'number:sum'],
        roundings: [],
        filters: [],
        expandedGroups: [],
        preventExpansion: false,
      })
    ).toMatchInlineSnapshot(`
      [
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "A")).Column1))",
          "aggregationResult": undefined,
          "children": [
            {
              "aggregationExpression": "sum((filter(TableName, TableName.Column1 == "A")).Column2)",
              "children": [],
              "elementType": "smartrow",
              "global": false,
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 4n,
                "s": 1n,
              },
            },
          ],
          "collapsible": true,
          "elementType": "group",
          "id": "/fcba17fd91a0ff43297ea7b8da8af0b3befbdc4266bb16c79b3b6554b133228d",
          "type": {
            "kind": "string",
          },
          "value": "A",
        },
        {
          "aggregationExpression": "count(unique((filter(TableName, TableName.Column1 == "B")).Column1))",
          "aggregationResult": undefined,
          "children": [
            {
              "aggregationExpression": "sum((filter(filter(TableName, TableName.Column1 == "B"), (filter(TableName, TableName.Column1 == "B")).Column2 == 2)).Column2)",
              "aggregationResult": {
                "type": {
                  "kind": "number",
                },
                "value": DeciNumber {
                  "d": 1n,
                  "infinite": false,
                  "n": 2n,
                  "s": 1n,
                },
              },
              "children": [],
              "collapsible": false,
              "elementType": "group",
              "id": "/1cdb9533372bfcf35d4e9b973a5d6f181ac200b6bfd7aba38051ee77ac5ac243/a60724359d845e8271fbda9708ee8cd10403c26bbe3c7775fc6fc855cb801170",
              "type": {
                "kind": "number",
                "unit": null,
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 2n,
                "s": 1n,
              },
            },
          ],
          "collapsible": false,
          "elementType": "group",
          "id": "/1cdb9533372bfcf35d4e9b973a5d6f181ac200b6bfd7aba38051ee77ac5ac243",
          "type": {
            "kind": "string",
          },
          "value": "B",
        },
        {
          "aggregationExpression": "count(unique(TableName.Column1))",
          "children": [
            {
              "aggregationExpression": "sum(TableName.Column2)",
              "children": [],
              "elementType": "smartrow",
              "global": true,
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 6n,
                "s": 1n,
              },
            },
          ],
          "elementType": "smartrow",
          "global": true,
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 2n,
            "s": 1n,
          },
        },
      ]
    `);
  });
});
