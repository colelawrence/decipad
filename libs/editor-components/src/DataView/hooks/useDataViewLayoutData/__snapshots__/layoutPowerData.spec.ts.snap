// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layoutPowerData > works with two column two groups tree and aggregation 1`] = `
{
  "meta": undefined,
  "type": {
    "columnNames": [
      "Column1",
      "Column2",
    ],
    "columnTypes": [
      {
        "kind": "string",
      },
      {
        "kind": "number",
        "unit": null,
      },
    ],
    "kind": "tree",
  },
  "value": Tree {
    "children": [
      Tree {
        "children": [
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 1n,
              "s": 1n,
            },
            "rootAggregation": {
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 1n,
                "s": 1n,
              },
            },
          },
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 3n,
              "s": 1n,
            },
            "rootAggregation": {
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 3n,
                "s": 1n,
              },
            },
          },
        ],
        "columns": [
          {
            "aggregation": {
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 4n,
                "s": 1n,
              },
            },
            "name": "Column2",
          },
        ],
        "originalCardinality": 2,
        "root": "A",
        "rootAggregation": {
          "meta": undefined,
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 1n,
            "s": 1n,
          },
        },
      },
      Tree {
        "children": [
          Tree {
            "children": [],
            "columns": [],
            "originalCardinality": 1,
            "root": DeciNumber {
              "d": 1n,
              "infinite": false,
              "n": 2n,
              "s": 1n,
            },
            "rootAggregation": {
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 2n,
                "s": 1n,
              },
            },
          },
        ],
        "columns": [
          {
            "aggregation": {
              "type": {
                "kind": "number",
              },
              "value": DeciNumber {
                "d": 1n,
                "infinite": false,
                "n": 2n,
                "s": 1n,
              },
            },
            "name": "Column2",
          },
        ],
        "originalCardinality": 1,
        "root": "B",
        "rootAggregation": {
          "meta": undefined,
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 1n,
            "s": 1n,
          },
        },
      },
    ],
    "columns": [
      {
        "aggregation": {
          "meta": undefined,
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 2n,
            "s": 1n,
          },
        },
        "name": "Column1",
      },
      {
        "aggregation": {
          "type": {
            "kind": "number",
          },
          "value": DeciNumber {
            "d": 1n,
            "infinite": false,
            "n": 6n,
            "s": 1n,
          },
        },
        "name": "Column2",
      },
    ],
    "originalCardinality": 1,
    "root": undefined,
    "rootAggregation": undefined,
  },
}
`;
